src/virt.rs

```diff
@@ -314,58 +314,150 @@ impl VirtContext {
         }
     }
 
-    pub fn handle_device_access_fault(&mut self, instr: &Instr, device: &VirtDevice) {
-        match instr {
-            Instr::Ld { rd, rs1, imm } => {
-                let address = utils::calculate_addr(self.get(rs1), *imm);
-                let offset = address - device.start_addr;
-                match device.device_interface.read_device(offset) {
-                    Ok(value) => {
-                        self.set(rd, value);
-                        self.pc += 4;
-                    }
-                    Err(err) => {
-                        panic!("Error reading {}: {}", device.name, err);
-                    }
-                }
-            }
-            Instr::CLd { rd, rs1, imm } => {
-                let address = self.get(rs1) + imm * 8;
-                let offset = address - device.start_addr;
-                match device.device_interface.read_device(offset) {
-                    Ok(value) => {
-                        self.set(rd, value);
-                        self.pc += 2;
-                    }
-                    Err(err) => {
-                        panic!("Error reading {}: {}", device.name, err);
-                    }
-                }
+    // TODO: think about passing instructions instead of registers
```

Comment:
Since all load/stores instuctions share registers insinde the same class (load or store), to my mind match won't be needed
Still, there probably isn't much to gain from that change, so registers are fine